src/lib.rs

```diff
@@ -0,0 +1,198 @@
+                self.routes.len() - 1
+            }
+
+        };
+
+        if path_nodes.len() == 1 {
+            return match &self.routes[index].value {
+                RouteValue::Value(_) => Err(SisoError { code: "E_VALUE_EXISTS".to_string() }),
+                RouteValue::None => {
+                    self.routes[index].value = value;
+                    Ok(true)
+                },
+            }
+        } else {
+            return self.routes[index].register(&path_nodes[1..], value);
+        }
+    }
+    fn find(&self, path_nodes : &[PathNode]) -> Result<RouteValue, SisoError> {
+        for a_route in self.routes.iter() {
+            if path_nodes[0] == a_route.node {
+                if path_nodes.len() == 1 {
+                    return Ok(a_route.value.clone());
+                } else {
+                    return a_route.find(&path_nodes[1..]);
+                }
+            }
+        }
+        Err(SisoError { code: "E_NOT_FOUND".to_string() })
+    }
+
+    fn new(node : PathNode, value : Option<RouteValue> ) -> Route {
+        Route { routes: Vec::new(), node, value: value.unwrap_or(RouteValue::None) }
+    }
+}
+
+// TODO: Allow generic value instead of strings only
+#[derive(Debug, Clone)]
+pub enum RouteValue {
+    Value(String),
+    None,
+
+}
+
+impl PartialEq for RouteValue {
+    fn eq(&self, other: &RouteValue) -> bool {
+        match self {
+            RouteValue::Value(self_value) => {
+                match other {
+                    RouteValue::Value(other_value) => self_value == other_value,
+                    RouteValue::None => false,
+                }
+            },
+            RouteValue::None => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SisoError {
+    code: String
+}
+
+impl fmt::Display for SisoError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "path already declared")
+    }
+}
+
+impl error::Error for SisoError {
+    fn description(&self) -> &str {
+        "path already declared"
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PathPattern {
+    name: String,
+    pattern: Regex,
+    // TODO: Add a way to parse the node and cast its value
+}
+
+#[derive(Debug, Clone)]
+pub enum PathNode {
+    PathPart(String),
+    PathPattern(PathPattern)
+}
+
+impl PartialEq for PathNode {
+    fn eq(&self, other: &PathNode) -> bool {
+        match self {
+            PathNode::PathPart(self_part) => {
+                match other {
+                    PathNode::PathPart(other_part) => self_part == other_part,
+                    _ => false
+                }
+            }
+            PathNode::PathPattern(self_pattern) => {
+                match other {
+                    PathNode::PathPattern(other_pattern) => self_pattern.name == other_pattern.name,
+                    _ => false
+                }
+            }
+        }
+    }
+}
```

Comment:
I think this is just `#[derive(PartialEq)]`?

Reply:
I guess so, did not know that it exited and thanks for the tip. That said, I've to figure out if it will work with the Regex type and the generic value I'll add to the PathPattern struct later (the end lib should returns the value for the path and the parsed path pattern, ie /article/:id should return the value of id casted to the generic type given by the path pattern). Will add a comment to rmind me that possibility and see if it works when the lib will be feature complete.